### PR TITLE
Amr/stepper motor

### PR DESCRIPTION
**Pull request by aransfor**
This is the stepper motors were using to control the angle of wave plates with an arduino

### CODE REVIEW

File: lib/servers/stepper_motor_server/stepper_motor_server.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            server = self.client[serServer]
+            ports = yield server.list_serial_ports()
+            if port not in ports:
+                continue
+            devName = '%s - %s' % (serServer, port)
+            devs += [(devName, (server, port))]
+        returnValue(devs)
+
+    @setting(101, 'move', steps='i')
+    def move(self, c, steps):
```
Comment (jayich) [inline lib/servers/stepper_motor_server/stepper_motor_server.py]: perhaps move_steps.  or step_move, or step.

File: lib/servers/stepper_motor_server/stepper_motor_server.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        p.write_line(code)
+        p.read_line()
+        ans = yield p.send()
+        returnValue(ans.read_line)
+
+
+class stepperMotorServer(DeviceServer):
+
+    deviceName = 'stepper_motor_server'
+    name = 'stepper_motor_server'
```
Comment (jayich) [inline lib/servers/stepper_motor_server/stepper_motor_server.py]: name = deviceName

File: lib/servers/stepper_motor_server/stepper_motor_server.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    def query(self, code):
+        """ Write, then read. """
+        p = self.packet()
+        p.write_line(code)
+        p.read_line()
+        ans = yield p.send()
+        returnValue(ans.read_line)
+
+
+class stepperMotorServer(DeviceServer):
```
Comment (jayich) [inline lib/servers/stepper_motor_server/stepper_motor_server.py]: StepperMotorServer

File: lib/servers/stepper_motor_server/stepper_motor_server.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        self.port = port
+        # The following opens a communication on a com port specifying a baudrate and a timeout
+        p = self.packet()
+        p.open(port)
+        p.baudrate(BAUDRATE)
+        p.read()  # clear out the read buffer
+        p.timeout(TIMEOUT)
+        yield p.send()
+
+    def packet(self):
```
Comment (jayich) [inline lib/servers/stepper_motor_server/stepper_motor_server.py]: from def packet to def query really feels like it should be inherited from a base serial class.

File: lib/servers/stepper_motor_server/stepper_motor_server.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
+# Created on Feb 12, 2017
+# @author: Anthony Ransford
+
+"""
+### BEGIN NODE INFO
+[info]
+name = stepper_motor_server
+version = 1.0
+description = drives stepper motors
+instancename = stepper_motor_server
```
Comment (jayich) [inline lib/servers/stepper_motor_server/stepper_motor_server.py]: probably good to be more specific in the server name.  I can easily see adding more stepper motor servers.

File: lib/servers/stepper_motor_server/stepper_motor_firmware/stepper_motor_firmware.ino [hunk 6]
```diff
@@ -0,0 +1,48 @@
+
```
Comment (jayich) [inline lib/servers/stepper_motor_server/stepper_motor_firmware/stepper_motor_firmware.ino]: I would just call the subfolder "firmware", you're already in the stepper_motor_server directory.

File: lib/clients/stepper_control/stepper_control.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
+from common.lib.clients.qtui.QCustomSpinBox import QCustomSpinBox
+from twisted.internet.defer import inlineCallbacks
+from PyQt4 import QtGui
+
+
+class stepper_control(QtGui.QWidget):
```
Comment (jayich) [inline lib/clients/stepper_control/stepper_control.py]: StepperControl

File: lib/clients/stepper_control/stepper_control.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        self.setLayout(layout)
+
+    def zero_scale(self, state):
+        self.angle = 0.0
+        self.degreewidget.spinLevel.setValue(0.0)
+
+    @inlineCallbacks
+    def change_angle(self, angle):
+        difference = angle - self.angle
+        steps = difference/1.8
```
Comment (jayich) [inline lib/clients/stepper_control/stepper_control.py]: self.degree_per_step = 1.8

I would put this in the server, and change_angle here just calls

server.change_angle(degree=value)

removing code from guis is great.

File: lib/servers/Example_Serial_Server.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+# @author: Anthony Ransford
+
+# The following is information for the node server in order to register
+# this server as available to the node. We generally use the same name
+# for instancename and name
+
+"""
+### BEGIN NODE INFO
+[info]
+name = Example_Serial_Server
```
Comment (jayich) [inline lib/servers/Example_Serial_Server.py]: What does this server do?

File: lib/servers/Example_Serial_Server.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        self.port = port
+        # The following opens a communication on a com port specifying a baudrate and a timeout
+        p = self.packet()
+        p.open(port)
+        p.baudrate(BAUDRATE)
+        p.read()  # clear out the read buffer
+        p.timeout(TIMEOUT)
+        yield p.send()
+
+    def packet(self):
```
Comment (jayich) [inline lib/servers/Example_Serial_Server.py]: this is repetition of the stepper code.  Were you intending to inherit this code for the stepper motor server?

File: lib/servers/Example_Serial_Server.py (hunk 9 above)
Comment (aransfor) [inline lib/servers/Example_Serial_Server.py]: this server does nothing, it is a template for future servers

File: lib/servers/Example_Serial_Server.py (hunk 10 above)
Comment (aransfor) [inline lib/servers/Example_Serial_Server.py]: no this is from the martinis group, basically there format has eliminated lots of bugs from serial and GPIB server lag so I am just sticking with it

File: lib/clients/stepper_control/stepper_control.py (hunk 8 above)
Comment (aransfor) [inline lib/clients/stepper_control/stepper_control.py]: I agree but I think the motor server should not know anything about angles, it just steps through. In the end I will most likely write a waveplate server that keeps track of and moves wave plates but in the interest of getting something usable this is a small amount of functionality in the GUI

Comment (jayich) [inline lib/clients/stepper_control/stepper_control.py]: I understand, this makes perfect sense.

File: lib/servers/stepper_motor_server/stepper_motor_server.py (hunk 4 above)
Comment (aransfor) [inline lib/servers/stepper_motor_server/stepper_motor_server.py]: I think this all should be part of the DeviceWrapper but for whatever reason this is not true. Perhaps someone should update labrad.devices and submit a PR

**Comment by aransfor** [comment]
@gregllong or @theoriginaljuice have a look over so I can push this through

**Comment by gregllong** [comment]
LGTM

**Pull request merged**